pkg/ring/lifecycler.go

```diff
@@ -246,18 +248,25 @@ func (i *Lifecycler) ChangeState(ctx context.Context, state IngesterState) error
 	return <-err
 }
 
-func (i *Lifecycler) getTokens() []uint32 {
+func (i *Lifecycler) getTokens() Tokens {
 	i.stateMtx.Lock()
 	defer i.stateMtx.Unlock()
+	if i.tokens == nil {
+		return nil
+	}
 	return i.tokens
 }
 
-func (i *Lifecycler) setTokens(tokens []uint32) {
+func (i *Lifecycler) setTokens(tokens Tokens) {
 	tokensOwned.WithLabelValues(i.RingName).Set(float64(len(tokens)))
 
 	i.stateMtx.Lock()
-	defer i.stateMtx.Unlock()
 	i.tokens = tokens
+	i.stateMtx.Unlock()
+
+	if err := i.tokens.StoreToFile(i.cfg.TokensFilePath); err != nil {
```

Comment:
I'm not sure it's a good idea accessing `i.tokens` outside the lock. Is it safe if we call `tokens.StoreToFile()` instead of `i.tokens.StoreToFile()`?

Reply:
There might be race in storing tokens in file in both cases. Best case to avoid this race of writing to files would be having it under the lock. Given that this happens during start-up, writing a small file inside the lock should be fine. WDYT?

Reply:
I see two possible race conditions:
1. You're accessing `i.tokens` outside the lock and another concurrent `setTokens()` may conflict
2. Two concurrent `setTokens()` will both try to store the file to disk and 1 will randomly win

I think the #2 is not a problem we should care because it's really remote and you use the move file operation in `StoreToFile`, so one will randomly win but the file integrity is guaranteed.

I believe #1 should be fixed. Given `i.tokens` is never manipulated (correct me if I'm wrong), only overwritten, it should be enough changing `i.tokens.StoreToFile()` to `tokens.StoreToFile()`.

Reply:
> 1 will randomly win

If the tokens written to the disk is not the winner, then is that not a problem? Because if we are reading the tokens after a restart of ingester, then the tokens that it read won't be all the tokens the is used.

Regarding point 1, yes, that should be fixed, but I think the order of file writes matter, or more like the tokens that are set at the end is should be the one reflected in the file. I will take another look at the functions calls to see if that is a problem.

(PS: using hash with a number will end up linking that PR/issue with this PR)

Reply:
Ok. Then let's be on the safest side (agree) and wrap the `StoreToFile` within the lock (you should just to rollback to `defer i.stateMtx.Unlock()`).